### PR TITLE
ARTEMIS-4419 Add federation support to AMQP broker connections

### PR DESCRIPTION
**Pull request by tabish121**
Allows federation of addresses and queues over an outbound AMQP broker connection and provide configuration via XML or broker propeties.

### CODE REVIEW

**Comment by jbertram** [comment]
This looks good!

It's not immediately clear to me how this compares to the existing federation functionality. Can you clarify that? Also, when would one choose one over the other? Does this make the existing federation obsolete?

**Comment by tabish121** [comment]
> This looks good!
> 
> It's not immediately clear to me how this compares to the existing federation functionality. Can you clarify that? Also, when would one choose one over the other? Does this make the existing federation obsolete?

This implementation isn't itself intended to be a replacement of the core based federation implementation as that would likely result in breakage of existing users of that federation which wouldn't fly in a minor version update.  This implementation exists on the AMQP broker connection feature to avoid breaking those existing users.  

Folks doing AMQP messaging might chose this implementation as it would then result in AMQP messages remaining in that form throughout their lifetime avoiding conversions to / from Core as they cross the core based federation consumer links.  

Another reason would be that unlike the core based implementation this implementation supports two way federation or simply "downstream" federation over the single outgoing connection avoiding the need to configure connection details for the remote to connect back to the local server which is how the core based version works for "downstream" configurations (which might not be possible in cloud based brokers, or might not be allowed due to firewall rules etc).

File: artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/AMQPBrokerConnection.java [hunk 1]
```diff
@@ -239,7 +271,44 @@ SimpleString getMirrorSNF(AMQPMirrorBrokerConnectionElement mirrorElement) {
...
+    * remote link closed handling should be ignored.
+    *
+    * @param id
+    *    A unique Id value that identifies the intercepter for later removal.
+    * @param intercepter
+    *    The predicate that will be called for any link close.
+    *
+    * @return this broker connection instance.
+    */
+   public AMQPBrokerConnection addLinkClosedIntercepter(String id, Predicate<Link> intercepter) {
```
Comment (gemmellr) [inline artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/AMQPBrokerConnection.java:286]: Intercepter -> Interceptor for consistency with other areas?

File: artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/federation/AMQPFederationPolicySupport.java [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+
+      if (element.getIncludes() != null && !element.getIncludes().isEmpty()) {
+         includes = new HashSet<>(element.getIncludes().size());
+
+         element.getIncludes().forEach(addressMatch -> includes.add(addressMatch.getAddressMatch()));
+      } else {
+         includes = Collections.EMPTY_SET;
+      }
+
+      if (element.getExcludes() != null && !element.getIncludes().isEmpty()) {
```
Comment (gemmellr) [inline artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/federation/AMQPFederationPolicySupport.java:457]: c&p error? null-checks excludes but then empty-checks includes.

File: artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/federation/AMQPFederationPolicySupport.java [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+      if (element.getIncludes() != null && !element.getIncludes().isEmpty()) {
+         includes = new HashSet<>(element.getIncludes().size());
+
+         element.getIncludes().forEach(queueMatch ->
+            includes.add(new AbstractMap.SimpleImmutableEntry<String, String>(queueMatch.getAddressMatch(), queueMatch.getQueueMatch())));
+      } else {
+         includes = Collections.EMPTY_SET;
+      }
+
+      if (element.getExcludes() != null && !element.getIncludes().isEmpty()) {
```
Comment (gemmellr) [inline artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/federation/AMQPFederationPolicySupport.java:511]: ditto

File: artemis-core-client/src/main/java/org/apache/activemq/artemis/core/config/FederationConfiguration.java [hunk 4]
```diff
@@ -74,7 +74,7 @@ public Map<String, FederationPolicy> getQueuePolicies() {
    }
 
    // strange spelling!, it allows a type match for singular of correct plural Policies from properties
-   public FederationConfiguration addQueuePolicie(FederationQueuePolicyConfiguration federationPolicy) {
+   public FederationConfiguration addQueuePolicy(FederationQueuePolicyConfiguration federationPolicy) {
```
Comment (gemmellr) [inline artemis-core-client/src/main/java/org/apache/activemq/artemis/core/config/FederationConfiguration.java:77]: Comment above method is inaccurate after the name changes.

File: artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/AMQPBrokerConnection.java (hunk 1 above)
Comment (tabish121) [reply artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/AMQPBrokerConnection.java]: fixed

File: artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/federation/AMQPFederationPolicySupport.java (hunk 2 above)
Comment (tabish121) [reply artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/federation/AMQPFederationPolicySupport.java]: fixed

File: artemis-core-client/src/main/java/org/apache/activemq/artemis/core/config/FederationConfiguration.java (hunk 4 above)
Comment (tabish121) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/core/config/FederationConfiguration.java:76]: removed the old comment

File: artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/federation/AMQPFederationPolicySupport.java (hunk 3 above)
Comment (tabish121) [reply artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/federation/AMQPFederationPolicySupport.java]: fixed

**Pull request merged**